MT-133929: initial port of the Audinate extclkin-gpt driver

Overdr0ne opened the pull request:
[MT-133929]

The driver provided from Audinate was very old. I can't tell exactly what version they were using, but they had code from <5.6 at least, and they also were not using the device tree. The device tree was introduced in 3.7. It was also built with buildroot, not using standard kernel module syntax, so that required some extra modification. This version compiles and instantiates the /dev/extclkin node, but I have not really tested it yet. I will make that a separate story, integrating with our particular use-case and getting everything working there.

Edit: I would also like to add here, there already exists an imx gpt driver in the kernel. Ideally Audinate would be using and configuring that. see drivers/clocksource/timer-imx-gpt.c. Without access to their source code though, we would surely completely break their interface trying to use that.

[MT-133929]: https://multitracks.atlassian.net/browse/MT-133929?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Overdr0ne committed:
extclkin-gpt: initial cleaned and formatted import of audinate driver

Overdr0ne committed:
extclkin-gpt: use newer class_create syntax

Overdr0ne committed:
extclkin-gpt: use replace deprecated ioremap_nocache function

Overdr0ne committed:
extclkin-gpt: use correct format specifier

Overdr0ne committed:
extclkin-gpt: fix module metadata

Overdr0ne commented:
Okay, hold of on merge, I was testing with this built as a module. After building into the kernel I encountered bugs. I need to load it later in the boot process...

Overdr0ne committed:
dts: mt-connect: cleanup whitespace

Overdr0ne committed:
extclkin-gpt: convert to platform device

Overdr0ne committed:
extclkin-gpt: cleanup whitespace

Overdr0ne committed:
extclkin-gpt: use correct device handle for debug

Overdr0ne committed:
dts: mt-connect: add entry for clkin_gpt

Overdr0ne committed:
config: mt-connect: extclkin-gpt need not be built into kernel

Overdr0ne committed:
extclkin-gpt: migrate extclkin from clk to clocksource

extclkin-gpt is not a clock, it is a timer that uses the LR clock as a
clocksource.

Overdr0ne commented:
I've cleaned things up, converted their driver to a platform driver and added necessary device tree entries. We are still having issues with DEP actually recognizing the device, but that seems to be on the DEP side. If now, I think that should be dealt with in a separate issue. I think this is ready to merge now.

AmeNote-Michael commented:
@Overdr0ne will test build with branch tonight. Want to confirm extclk providing correct output.

AmeNote-Michael commented:
@Overdr0ne as discussed in channel, the extclk is currently not working properly likely due to not having counter configured properly.

`         71075489250	                   0
         71075500375	                   0
         71075511750	                   0
         71075522250	                   0
         71075533500	                   0
         71075544625	                   0
         71075555000	                   0
`

Overdr0ne committed:
dts: mt-connect: add fixed fbclk and pinmux

Overdr0ne committed:
dts: mt-connect: move pinctrl to clkin_gpt node for now

It's not necessarily wrong to do it this way, but it is more
configurable to have the pinctrl in the clock node, because then you
can just change the clock without touching the pin config. I am still
investigating why it didn't work when pinctrl was on the clock node. I
don't see anything else configuring that pin, but maybe I missed
something, I'm just not sure...

Overdr0ne commented:
As Michael pointed out, after muxing the feedback clock, the driver seems to be getting the correct GPT count. I have added that change, though not to the location that I would have preferred. For some reason, pinmuxing from the fixed fbclk node is not working. Could be something external is configuring that pin. More investigation required.

AmeNote-Michael commented:
@Overdr0ne Sam, create ticket to re-review this dts setup issue, but set low priority. Copy ticket link here.

I am building with this branch to test.

AmeNote-Michael commented:
I can confirm that the correct configuration exists such that extclkin now gives correct information.

`MultiTracks Platform (Development) 2.1.4 multitracks ttymxc1

multitracks login: root
multitracks:~# cd /dev 
multitracks:/dev# cat extclkin 
         28095616375	         18183301400
         28095675250	         18183360000
         28095688500	         18183373200
         28095699500	         18183384200
`

Overdr0ne commented:
Issue created for clock pinmuxing issue:
https://multitracks.atlassian.net/browse/MT-135825?atlOrigin=eyJpIjoiZmRhMGIwZDFmMTlmNDcxZWExOTMwNDI2M2MxZTczOGEiLCJwIjoiaiJ9
